### PR TITLE
Allow overriding functions env vars with .env file

### PR DESCRIPTION
**Pull request by RaeesBhatti**
**- Summary**
Fixes: https://github.com/netlify/cli/issues/734

**- Test plan**
Env variables set in [app.netlify.com](https://app.netlify.com) override variables in `.env` files for functions when running with `netlify dev`.
**- If the current behavior is a bug, please provide the steps to reproduce.**
* Link your project to a Netlify site using `netlify link`
* Create an env variable in [app.netlify.com](https://app.netlify.com) for your site
* Create a `.env` file in your function directory with the same variable name (access that env variable in your function)
* Run `netilfy dev` and access your function, see the value for the env var

**- Description for the changelog**
* Record `process.env` before creating functions handler
* Provide variables from `.env` file to `lambda-local` (will result in overriding `process.env`)
* Restore `process.env` from earlier recorded object after `lambda-local` has finished executing

**- A picture of a cute animal (not mandatory but encouraged)**
🦁

### CODE REVIEW

File: src/utils/serve-functions.js [hunk 1]
```diff
@@ -154,14 +156,18 @@ function createHandler(dir) {
 
     const callback = createCallback(response)
     // we already checked that it exports a function named handler above
+    let envConfig = {}
+    if (fs.existsSync(path.resolve(path.dirname(functionPath)))) {
+      envConfig = dotenv.parse(fs.readFileSync(path.resolve(path.dirname(functionPath), '.env')))
```
Comment (ehmicky) [inline src/utils/serve-functions.js]: If the `.env` file does not exist, `fs.readFileSync()` will throw. Is this ok?

Comment (RaeesBhatti) [reply src/utils/serve-functions.js]: We're checking for `fs.fs.existsSync` first and then calling `fs.readFileSync`

Comment (ehmicky) [reply src/utils/serve-functions.js]: We are checking if `{functionPath}` exists but not if `{functionPath}/.env` does.

Comment (RaeesBhatti) [reply src/utils/serve-functions.js]: Oh good catch! Thanks

**Comment by RaeesBhatti** [comment]
Actually on the second note, I made sure that the app server env variables are [not set on the main process](https://github.com/netlify/cli/blob/57e49745adb64ec0d6b59b18d4f5cc167faccb23/src/commands/dev/index.js#L320-L328), they're instead only set on the child process on app server.

Similarly in the function server, since we're not creating a child process to serve functions, I record the main process variables [before any function is run](https://github.com/netlify/cli/pull/735/files#diff-cf2bd94d7688f0711bbd931e6c5e0397R199-R200), run the function with [its specific env vars](https://github.com/netlify/cli/pull/735/files#diff-cf2bd94d7688f0711bbd931e6c5e0397R159-R170), then [destroy those functions specific vars](https://github.com/netlify/cli/pull/735/files#diff-cf2bd94d7688f0711bbd931e6c5e0397R171) after function execution is complete and [reset the vars from main process](https://github.com/netlify/cli/pull/735/files#diff-cf2bd94d7688f0711bbd931e6c5e0397R206-R207).

On the other hand, if app server or any of the functions is compromised, they have access to file system and can read anything.

**Comment by erezrokah** [comment]
@RaeesBhatti, sorry I should have been more clear with my note. It wasn't related to the cli itself, was referring to deployed functions in general.
Each site function inherits all env. variables configured for that site and not just the ones it should have access to.

**Comment by RaeesBhatti** [comment]
Ah I see. I think we implemented the wrong behavior in the CLI then.

**Comment by erezrokah** [comment]
I thought this is done here https://github.com/netlify/cli/blob/0d076ac7f02fe296ea2623cbe26496fd506bebcf/src/commands/dev/exec.js#L17
https://github.com/netlify/cli/blob/0d076ac7f02fe296ea2623cbe26496fd506bebcf/src/utils/dev.js#L82

**Comment by RaeesBhatti** [comment]
Yes that adds env variables from the production to Netlify Dev, this PR is specifically about supporting `.env` files for Netlify Dev specific functions. e.g. If you want to override prod env vars.

**Comment by erezrokah** [comment]
I think we're agreeing on the same thing :) I don't think I under your comment about implementing the wrong thing. Looks like that with your fix the CLI behaves like it should - overriding prod env with ones from `.env` files

**Pull request merged**